File: web/test/commentHistory.test.js
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { remember, suggestions } from '../domain/commentHistory.js';

test('remember trims and skips empty', () => {
  const existing = ['Fill-up'];
  assert.deepEqual(remember(existing, '   '), existing);
  assert.deepEqual(remember(existing, null), existing);
  assert.deepEqual(remember(existing, '  Market run  '), ['Market run', 'Fill-up']);
});

test('remember merges case insensitive keeping latest casing', () => {
  assert.deepEqual(remember(['fill-up', 'Groceries'], 'Fill-up'), ['Fill-up', 'Groceries']);
});

test('remember moves to front and caps at fifty', () => {
  const existing = [];
  for (let i = 0; i < 50; i++) existing.push(`Note ${i}`);
  const next = remember(existing, 'Newest');
  assert.equal(next.length, 50);
  assert.equal(next[0], 'Newest');
  assert.equal(next[1], 'Note 0');
  assert.equal(next[49], 'Note 48');
});

test('suggestions prefix then contains recency', () => {
  const comments = ['Fill-up', 'Fun night', 'Farmer market', 'Coffee'];
  const matches = suggestions(comments, 'f');
  assert.deepEqual(matches, ['Fill-up', 'Fun night', 'Farmer market', 'Coffee']);
  assert.deepEqual(matches.slice(0, 3), ['Fill-up', 'Fun night', 'Farmer market']);
});

test('suggestions empty query or no match', () => {
  const comments = ['Fill-up', 'Coffee'];
  assert.deepEqual(suggestions(comments, ''), []);
  assert.deepEqual(suggestions(comments, '   '), []);
  assert.deepEqual(suggestions(comments, 'zzz'), []);
});

File: app/src/test/java/com/example/envelopemoney/CommentHistoryTest.java
package com.example.envelopemoney;

import org.junit.Test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

public class CommentHistoryTest {

    @Test
    public void remember_trimsAndSkipsEmpty() {
        List<String> existing = Collections.singletonList("Fill-up");
        assertEquals(existing, CommentHistory.remember(existing, "   "));
        assertEquals(existing, CommentHistory.remember(existing, null));
        assertEquals(Arrays.asList("Market run", "Fill-up"),
                CommentHistory.remember(existing, "  Market run  "));
    }

    @Test
    public void remember_mergesCaseInsensitiveKeepingLatestCasing() {
        List<String> existing = new ArrayList<>(Arrays.asList("fill-up", "Groceries"));
        List<String> next = CommentHistory.remember(existing, "Fill-up");
        assertEquals(Arrays.asList("Fill-up", "Groceries"), next);
    }

    @Test
    public void remember_movesToFrontAndCapsAtFifty() {
        List<String> existing = new ArrayList<>();
        for (int i = 0; i < 50; i++) {
            existing.add("Note " + i);
        }
        List<String> next = CommentHistory.remember(existing, "Newest");
        assertEquals(50, next.size());
        assertEquals("Newest", next.get(0));
        assertEquals("Note 0", next.get(1));
        assertEquals("Note 48", next.get(49));
    }

    @Test
    public void suggestions_prefixThenContainsRecency() {
        List<String> comments = Arrays.asList("Fill-up", "Fun night", "Farmer market", "Coffee");
        List<String> matches = CommentHistory.suggestions(comments, "f");
        assertEquals(Arrays.asList("Fill-up", "Fun night", "Farmer market", "Coffee"), matches);
        assertEquals(Arrays.asList("Fill-up", "Fun night", "Farmer market"),
                CommentHistory.suggestions(comments, "f").subList(0, 3));
    }

    @Test
    public void suggestions_emptyQueryOrNoMatch() {
        List<String> comments = Arrays.asList("Fill-up", "Coffee");
        assertTrue(CommentHistory.suggestions(comments, "").isEmpty());
        assertTrue(CommentHistory.suggestions(comments, "   ").isEmpty());
        assertTrue(CommentHistory.suggestions(comments, "zzz").isEmpty());
    }
}
